Lib/multiprocessing/shared_memory.py

```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
 
     def close(self):
-        self.buf.release()
-        self._mmap.close()
+        if self._buf is not None:
+            self._buf.release()
+            self._buf = None
+        if self._mmap is not None:
+            self._mmap.close()
+            self._mmap = None
 
     def unlink(self):
         """Windows ensures that destruction of the last reference to this
         named shared memory block will result in the release of this memory."""
         pass
 
 
-class PosixSharedMemory(_PosixSharedMemory):
+# FreeBSD (and perhaps other BSDs) limit names to 14 characters.
+_SHM_SAFE_NAME_LENGTH = 14
+
+# shared object name prefix
+if os.name == "nt":
+    _SHM_NAME_PREFIX = 'wnsm_'
+else:
+    _SHM_NAME_PREFIX = '/psm_'
+
+
+def _make_filename():
+    """Create a random filename for the shared memory object.
+    """
+    # number of random bytes to use for name
+    nbytes = (_SHM_SAFE_NAME_LENGTH - len(_SHM_NAME_PREFIX)) // 2
+    assert nbytes >= 2, '_SHM_NAME_PREFIX too long'
+    name = _SHM_NAME_PREFIX + secrets.token_hex(nbytes)
+    assert len(name) <= _SHM_SAFE_NAME_LENGTH
+    return name
+
+
+class PosixSharedMemory:
+
+    # defaults so close() and unlink() can run without errors
+    fd = -1
+    name = None
+    _mmap = None
+    _buf = None
 
     def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
         if name and (flags is None):
-            _PosixSharedMemory.__init__(self, name, mode=mode)
+            flags = 0
         else:
-            if name is None:
-                name = f'psm_{os.getpid()}_{random.randrange(100000)}'
             flags = O_CREX if flags is None else flags
-            _PosixSharedMemory.__init__(
-                self,
-                name,
-                flags=flags,
-                mode=mode,
-                size=size,
-                read_only=read_only
-            )
-
+        if flags & O_EXCL and not flags & O_CREAT:
+            raise ValueError("O_EXCL must be combined with O_CREAT")
+        if name is None and not flags & O_EXCL:
+            raise ValueError("'name' can only be None if O_EXCL is set")
+        flags |= os.O_RDONLY if read_only else os.O_RDWR
+        self.flags = flags
+        self.mode = mode
+        if not size >= 0:
+            raise ValueError("'size' must be a positive integer")
+        if name is None:
+            self._open_retry()
+        else:
+            self.name = name
+            self.fd = _posixshmem.shm_open(name, flags, mode=mode)
+        if size:
+            try:
+                os.ftruncate(self.fd, size)
+            except OSError:
+                self.unlink()
+                raise
         self._mmap = mmap.mmap(self.fd, self.size)
-        self.buf = memoryview(self._mmap)
+        self._buf = memoryview(self._mmap)
+
+    @property
+    def size(self):
+        "Size in bytes."
+        if self.fd >= 0:
+            return os.fstat(self.fd).st_size
+        else:
+            return 0
```

Comment:
Mmm... since you make a `os.fstat` call this may fail. And it's kinda unusual for a `getattr` to fail. Also, AFAIU the size may change if the memory is enlarged (it can happen I suppose). As such I think it may make sense to provide this as a method, not a property.

Reply:
In commit 0d3d06f7cbcb47b2dee6c027c35265795052c899, I have eliminated the use of ``os.fstat()`` in the property for ``size``.  Instead, it is now an unchanging value stored as part of the ``__init__()``.